contracts/ConsensusApp.sol

```diff
@@ -4,237 +4,56 @@ pragma experimental ABIEncoderV2;
-        uint256 numParticipants
-    ) internal pure returns (bool) {
-        if (furtherVotesRequiredInitialized(newCommitmentData, numParticipants)) {
-            validateBalancesUnchanged(oldCommitmentData, newCommitmentData);
-            return true;
-        } else {
-            return false;
-        }
-    }
-
-    function validVote(
-        ConsensusCommitmentData memory oldCommitmentData,
-        ConsensusCommitmentData memory newCommitmentData
-    ) internal pure returns (bool) {
-        if (
-            oldCommitmentData.furtherVotesRequired > 1 &&
-            furtherVotesRequiredDecremented(oldCommitmentData, newCommitmentData)
-        ) {
-            validateBalancesUnchanged(oldCommitmentData, newCommitmentData);
-            validateProposalsUnchanged(oldCommitmentData, newCommitmentData);
-            return true;
-        } else {
-            return false;
-        }
-    }
-
-    function validVeto(
-        ConsensusCommitmentData memory oldCommitmentData,
-        ConsensusCommitmentData memory newCommitmentData
-    ) internal pure returns (bool) {
-        if (
-            oldCommitmentData.furtherVotesRequired > 0 &&
-            newCommitmentData.furtherVotesRequired == 0 &&
-            balancesUnchanged(oldCommitmentData, newCommitmentData)
-        ) {
-            return true;
-        } else {
-            return false;
-        }
-    }
-
-    function validFinalVote(
-        ConsensusCommitmentData memory oldCommitmentData,
-        ConsensusCommitmentData memory newCommitmentData
-    ) internal pure returns (bool) {
-        if (
-            oldCommitmentData.furtherVotesRequired == 1 &&
-            newCommitmentData.furtherVotesRequired == 0 &&
-            balancesUpdated(oldCommitmentData, newCommitmentData)
-        ) {
-            return true;
-        } else {
-            return false;
-        }
-    }
 
-    function validPass(
-        ConsensusCommitmentData memory oldCommitmentData,
-        ConsensusCommitmentData memory newCommitmentData
-    ) internal pure returns (bool) {
-        if (
-            oldCommitmentData.furtherVotesRequired == 0 &&
-            newCommitmentData.furtherVotesRequired == 0
-        ) {
-            validateBalancesUnchanged(oldCommitmentData, newCommitmentData);
+        ConsensusAppData memory appDataA = appData(a.appData);
+        ConsensusAppData memory appDataB = appData(b.appData);
+
+        if (identical(a.outcome, b.outcome)) {
+            if (appDataB.furtherVotesRequired == numParticipants - 1) {
+                // propose
+                require(appDataA.furtherVotesRequired == 0, 'ConsensusApp: invalid propose, furtherVotesRequired must be transition from zero');
+                return true;
+            } else if (appDataB.furtherVotesRequired == 0) {
+                // veto or pass
+                require(appDataB.proposedOutcome.length == 0, 'ConsensusApp: invalid veto or invalid pass, proposeOutcome must transition to null');
```

Comment:
There is no concept of `null` in Solidity, so this wording might be a bit confusing. Perhaps `empty` is a better term?

Reply:
```suggestion
                require(appDataB.proposedOutcome.length == 0, 'ConsensusApp: invalid veto or invalid pass, proposedOutcome must transition to empty');
```